include/oneapi/dpl/pstl/hetero/dpcpp/sycl_iterator.h

```diff
@@ -213,18 +234,20 @@ __internal::sycl_iterator<access_mode::discard_read_write, T, Allocator>
 }
 
 // end
-template <typename T, typename Allocator, access_mode Mode>
-__internal::sycl_iterator<Mode, T, Allocator> end(sycl::buffer<T, /*dim=*/1, Allocator> buf, sycl::mode_tag_t<Mode>)
+template <typename T, typename Allocator, typename ModeTagT>
+__internal::sycl_iterator<__internal::__access_mode_resolver<ModeTagT>::__value, T, Allocator>
+end(sycl::buffer<T, /*dim=*/1, Allocator> buf, ModeTagT)
 {
-    return __internal::sycl_iterator<Mode, T, Allocator>{buf, __dpl_sycl::__get_buffer_size(buf)};
+    return __internal::sycl_iterator<__internal::__access_mode_resolver<ModeTagT>::__value, T, Allocator>{
+        buf, __dpl_sycl::__get_buffer_size(buf)};
 }
 
-template <typename T, typename Allocator, access_mode Mode>
-__internal::sycl_iterator<__internal::_ModeConverter<Mode>::__value, T, Allocator>
-    end(sycl::buffer<T, /*dim=*/1, Allocator> buf, sycl::mode_tag_t<Mode>, __dpl_sycl::__no_init)
+template <typename T, typename Allocator, typename ModeTagT>
+__internal::sycl_iterator<__internal::__access_mode_resolver<ModeTagT, __dpl_sycl::__no_init>::__value, T, Allocator>
+end(sycl::buffer<T, /*dim=*/1, Allocator> buf, ModeTagT, __dpl_sycl::__no_init)
```

Comment:
Am I correct in understanding that this should not break any existing code which would pass in `sycl::mode_tag_t<Mode>` to specify an access mode (without host_task stuff)?

Is it just that the sycl spec leaves the types of deduction tags unspecified and we should not use the internals in our code either?

Reply:
> Am I correct in understanding that this should not break any existing code which would pass in sycl::mode_tag_t<Mode> to specify an access mode (without host_task stuff)?

You are right, it allows passing `sycl::mode_tag_t` as any other implementation-defined deduction tag type.

> Is it just that the sycl spec leaves the types of deduction tags unspecified and we should not use the internals in our code either?

This is also correct. The types of the [SYCL 2020 deduction tags](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_deduction_tags) are unspecified, and we do not want to rely on the internals of the specific implementations.